3836/3836.md

```diff
@@ -0,0 +1,666 @@
+_ -> error "decodeTerm: impossible happened!",
+```
+
+The error is triggered when a `peerSharing` value, in a context in which it is
+guarded to be between 0 and 2 (`| ..., peerSharing >= 0, peerSharing <= 2`),
+pattern-matches to a number other than 0, 1, or 2. Failure of the primary guard
+already triggers failure via `Left`, so I've eliminated the mention of `error`
+by reorganized the pattern-matching while preserving the existing circumstances
+in which failure via `Left` is indicated.
+
+**Fix: refactor**
+#fixed
+
+
+### 14: `ouroboros-network-api/src/Ouroboros/Network/AnchoredFragment.hs:165`
+```hs
+anchorFromPoint GenesisPoint _     = error "anchorFromPoint: genesis point"
+```
+
+I don't fully understand, logically, why calling `anchorFromPoint` on
+`GenesisPoint` warrants an error. It seems like the natural result would be
+`AnchorGenesis`, an `Anchor` constructor seemingly for this purpose, rather than
+an error. Such a change would bring `anchorFromPoint` and `anchorToPoint` closer
+to being logical inverses of one another, if this is desirable.
+
+I note that this function is exported but does not appear to be used elsewhere
+within the IOG org (per an org-wide Github search). Perhaps it could be marked
+deprecated and/or removed?
+
+**Fix: uncertain**
+#unfixed
+
+#input-desired
+
+
+### 15: `ouroboros-network-framework/src/Ouroboros/Network/Protocol/Handshake/Version.hs:73`
+```hs
+[] -> error "foldMapVersions: precondition violated"
+```
+
+This error appears in `foldMapVersions`, which converts a collection of things
+into a collection of `Version`s (represented as a `Versions` value). The
+precondition in question is that the collection of things be nonempty. The
+rationale for this precondition is not obvious - I do not immediately see why it
+is erroneous to have an empty `Versions` value, which can be constructed
+manually based on this module's exports. This module might benefit from
+documentation of that rationale, and/or restriction of its exports to prevent
+manual construction of empty `Versions` values.
+
+Separately, `foldMapVersions` could be modified to save a `toList` conversion:
+```hs
+-- Existing:
+foldMapVersions f fx = case toList fx of
+    [] -> error "foldMapVersions: precondition violated"
+    xs -> foldl1 (<>) (map f xs)
+
+-- Proposed:
+foldMapVersions f fx
+  | null fx = error "foldMapVersions: precondition violated"
+  | otherwise = Versions (foldMap' f' fx)
+  where
+    f' = getVersions . f
+```
+
+This would better fit with `Foldable` idioms - `toList` advises: "[i]f the
+entire list is intended to be reduced via a fold, just fold the structure
+directly bypassing the list." This change would affect strictness: `foldMap'` is
+left-associative and strict in its accumulator, while `foldl1` is
+left-associative but lazy in its accumulator. In principle, this should be a net
+gain for efficiency, since lazy left-folds are generally undesirable.
+
+**Fix: document**
+#unfixed
+
+#input-desired
```

Comment:
This is a precondition; passing an empty list would mean that the handshake will always fail since one doesn't offer any choice to the remote side.

I agree that we should just use `null` instead, although I would still use `foldMap' f fx` and relay on the `Versions` semigroup instance - this is a good example how to use it.

Alternatively, we could use non-empty lists and [`foldMap1`](https://hackage.haskell.org/package/foldable1-classes-compat-0.1/docs/Data-Foldable1.html#v:foldMap1-39-).  `foldMap1` has been introduced to `base-4.18`.  I hope soon we can switch to `ghc-9.6` (`base-4.18`) and forget about `ghc-8.10`.

Reply:
I don't think we can use `foldMap` or `foldMap'` and at the same time rely on the `Semigroup Versions` instance with `f`. `foldMap` and `foldMap'` rely on a `Monoid` instance, which `Versions` doesn't (and, as I can now see, can't) have, so we'd need to deconstruct it and rely on the `Semigroup Map` instance with `f'` instead.

Maybe I'll leave things as-is for now, but record an issue referring to this code and this discussion that can be revisited when y'all switch to 9.6. How does that sound?

Reply:
I've opened #4677  for this.